python/cudf/cudf/_lib/transform.pyx

```diff
@@ -156,3 +162,30 @@ def one_hot_encode(Column input_column, Column categories):
     )
 
     return encodings
+
+
+def compute_column(list columns, tuple column_names, expr: str):
+    """Compute a new column by evaluating an expression on a set of columns.
+
+    Parameters
+    ----------
+    df : cudf.DataFrame
+        The DataFrame that we are applying functions to.
+    expr : str
+        The expression to evaluate. Must be a single expression (not a
+        statement, i.e. no assignment).
+    """
+    visitor = parse_expression(expr, column_names)
+
+    # At the end, all the stack contains is the expression to evaluate.
+    cdef Expression cudf_expr = visitor.expression
```

Comment:
Are we ever going to not want to do both these things in sequence? i.e., should `parse_expression` instead just return an `Expression`?

Reply:
The reason that we need to do it this way is because we need the visitor to keep all the child nodes alive. An `Expression` doesn't own its subnodes, so they'll be garbage collected as soon as they go out of scope. This mimics the C++ structure where nodes don't own the nodes they refer to either.

Reply:
Couldn't this be abstracted away a bit by having the `Expression` hold the `visitor` as a Python attribute?

Reply:
There are certainly ways that we could fix this by adding in extra ownership linkages, but they all feel less natural than the current code structure. The visitor is essentially a parser for an expression, so it makes intuitive sense to me that it owns the result of parsing. Conversely, why would an `Expression` own the parse that was used to generate it? Alternatively, a more natural solution IMO would be to add ownership linkages between all the different type of `Expression`, but that would be an explicit break from the C++ design and could cause different problems.

Reply:
The issue is having a Python type that has C++ "ownership' semantics. An analogue is RMM's device buffer, which needs the underlying memory resource kept alive. In the Cython wrappers, we achieve this by having it hold a reference to it.

> Conversely, why would an Expression own the parse that was used to generate it?

Because it needs it alive? Ownership semantics should be abstracted away for Python types as much as possible. It's not something the consumer of a Python API should have to worry about.

Reply:
I'm happy to defer the discussion and possible implementation to a follow-up, going to approve!

Reply:
I'm going to defer this discussion. I think this question reflects some design decisions in the C++ layer and I'd rather come to a resolution there first before deciding on any approach here.

Reply:
To be clear, I don't think this is a question of Python vs C++ semantics. I think this a question of what the design of libcudf's expression evaluation looks like.

Reply:
Opened #10744 to discuss the underlying issues further.